contrib/src/main/java/com/datatorrent/contrib/nifi/AbstractNiFiOutputOperator.java

```diff
@@ -0,0 +1,118 @@
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package com.datatorrent.contrib.nifi;
+
+import java.io.IOException;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.apache.nifi.remote.Transaction;
+import org.apache.nifi.remote.TransferDirection;
+import org.apache.nifi.remote.client.SiteToSiteClient;
+
+import com.datatorrent.api.Context;
+import com.datatorrent.common.util.BaseOperator;
+import com.datatorrent.netlet.util.DTThrowable;
+
+/**
+ * This is the base implementation of a NiFi output operator.&nbsp;
+ * A concrete operator should be created from this skeleton implementation.
+ * <p>
+ * <br>
+ * Ports:<br>
+ * <b>Input</b>: Can have any number of input ports<br>
+ * <b>Output</b>: no output port<br>
+ * <br>
+ * Properties:<br>
+ * None<br>
+ * <br>
+ * Compile time checks:<br>
+ * None<br>
+ * <br>
+ * Run time checks:<br>
+ * None<br>
+ * <br>
+ * Benchmarks:<br>
+ * TBD<br>
+ * <br>
+ * </p>
+ *
+ * @displayName Abstract NiFi Output
+ * @category Messaging
+ * @tags output operator
+ *
+ * @since 3.3.0
+ *
+ */
+public abstract class AbstractNiFiOutputOperator<T> extends BaseOperator
+{
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(AbstractNiFiOutputOperator.class);
+
+  protected SiteToSiteClient client;
+  protected final SiteToSiteClient.Builder siteToSiteBuilder;
+  protected final NiFiDataPacketBuilder<T> dataPacketBuilder;
+
+  /**
+   * @param siteToSiteBuilder the builder for a NiFi SiteToSiteClient
+   * @param dataPacketBuilder a builder to produce NiFiDataPackets from incoming data
+   */
+  public AbstractNiFiOutputOperator(final SiteToSiteClient.Builder siteToSiteBuilder,
+      final NiFiDataPacketBuilder<T> dataPacketBuilder)
+  {
+    this.siteToSiteBuilder = siteToSiteBuilder;
+    this.dataPacketBuilder = dataPacketBuilder;
+  }
+
+  @Override
+  public void setup(Context.OperatorContext context)
+  {
+    this.client = siteToSiteBuilder.build();
+  }
+
+  @Override
+  public void teardown()
+  {
+    try {
+      client.close();
+    } catch (IOException e) {
+      LOGGER.error(e.getMessage(), e);
+    }
+  }
+
+  protected void processTuple(T tuple)
+  {
+    final NiFiDataPacket niFiDataPacket = dataPacketBuilder.createNiFiDataPacket(tuple);
+
+    try {
+      final Transaction transaction = client.createTransaction(TransferDirection.SEND);
```

Comment:
If creating a transaction is costly operation we might create batch of tuples and send them in one transaction during "endWindow". Creating a transaction to send only one tuple is optimal?